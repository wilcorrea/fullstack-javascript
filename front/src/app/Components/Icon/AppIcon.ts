import { CreateElement, RenderContext } from 'vue'

/**
 * @type {string}
 */
const block = 'app-icon'

/**
 * @component {AppIcon}
 */
export default {
  /**
   */
  functional: true,
  /**
   */
  name: 'AppIcon',
  /**
   * @param {CreateElement} h
   * @param {RenderContext} context
   */
  render (h: CreateElement, context: RenderContext) {
    const { props, listeners } = context
    const style = {}
    if (props.color) {
      // @ts-ignore
      style['color'] = props.color
    }
    if (listeners.click) {
      // @ts-ignore
      style['cursor'] = 'pointer'
    }
    const data = { class: `${block} material-icons`, style, on: { ...listeners } }
    let name = props.name
    if (!name) {
      name = 'warning'
      console.warn('[app-icon]', ' ', 'using a fallback icon, please provide a name')
    }

    return h('span', data, name)
  }
}
