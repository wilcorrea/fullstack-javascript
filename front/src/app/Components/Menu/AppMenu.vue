<template>
  <div class="AppMenu">
    <div
      class="AppMenu__backdrop"
      :class="{'AppMenu__trigger--visible': showMenu}"
      @click="appMenuToggle"
    />
    <div
      class="AppMenu__trigger"
      :class="{'AppMenu__trigger--triggered': showMenu}"
      @click="appMenuToggle"
    />
    <div
      class="AppMenu__links"
      :class="{'AppMenu__links--open': showMenu}"
    >
      <div class="AppMenu__closer">
        <app-icon
          name="close"
          color="white"
          @click="appMenuToggle"
        />
      </div>
      <ul>
        <li
          v-for="(link, index) in links"
          :key="index"
          @click="appMenuToggle"
        >
          <router-link :to="link.path">
            {{ link.label }}
          </router-link>
        </li>
      </ul>
    </div>
  </div>
</template>

<script lang="js">
export default {
  /**
   */
  name: 'AppMenu',
  /**
   */
  props: {
    links: {
      type: Array,
      required: true
    }
  },
  /**
   */
  data: () => ({
    showMenu: false
  }),
  /**
   */
  methods: {
    /**
     */
    appMenuToggle () {
      this.showMenu = !this.showMenu
    }
  }
}
</script>

<style
  lang="stylus"
  scoped
>
@import "~src/assets/variables.styl"

.AppMenu__absolute
  position absolute
  top 0
  left 0

.AppMenu__transition
  position absolute
  background $primary
  transition-property transform, opacity
  transition-timing-function ease-in-out
  transition-duration .3s

.AppMenu
  @extend .AppMenu__absolute
  > .AppMenu__backdrop
    @extend .AppMenu__absolute
    @extend .AppMenu__transition
    z-index 1
    height 100vh
    width 100vw
    background-color rgba(0, 0, 0, 0.45)
    transition opacity .5s ease-in-out, transform .1s .5s
    transform translateX(-110vw)
    opacity 0

    &.AppMenu__trigger--visible
      transition opacity .5s ease-in-out, transform .1s .0s ease-in-out
      transform translateX(0)
      opacity 1

  > .AppMenu__trigger
    @extend .AppMenu__transition
    z-index 2
    box-shadow 3px 1px 9px 4px rgba(0, 0, 0, 0.2)
    width 80px
    height 80px
    border-radius 40px
    top -40px
    left -40px
    cursor pointer
    opacity 1
    animation shadow-pulse 2s infinite

    &:hover
      transform scale(2)

    &.AppMenu__trigger--triggered
      opacity 0

  > .AppMenu__links
    @extend .AppMenu__transition
    z-index 3
    box-shadow 3px 1px 7px 4px rgba(0, 0, 0, 0.355)
    padding 20px
    height 100vh
    width 200px
    transform translateX(-250px)

    &.AppMenu__links--open
      transform translateX(0)

    > .AppMenu__closer
      float right

    > ul > li > a
      text-decoration none
      display block
      margin 30px 0
      color #ffffff
      font-size 1.6em

  @keyframes shadow-pulse
  {
    0% {
      box-shadow 3px 1px 9px 4px rgba(0, 0, 0, 0.2), 0 0 0 0 rgba(0, 0, 0, 0.2)
    }
    100% {
      box-shadow 3px 1px 9px 4px rgba(0, 0, 0, 0.2), 0 0 0 35px rgba(0, 0, 0, 0)
    }
  }
</style>
