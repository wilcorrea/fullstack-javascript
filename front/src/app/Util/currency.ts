/**
 * @param value
 * @param {currency} precision
 * @param {string} thousands
 * @param {string} decimal
 * @param {string} currency
 * @returns {string}
 */
export const money = (
  value: number,
  precision: number = 2,
  thousands: string = '.',
  decimal: string = ',',
  currency: string = 'R$ '
): string => {
  const pieces = Number(value).toFixed(precision).split('.')
  const left: any = pieces[0]
  const right = (pieces.length === 1 ? '' : pieces[1]).padEnd(precision, '0')
  // noinspection RegExpRedundantEscape
  const answer = left
    .reverse()
    .replace(/(\d{3})/g, `$1${thousands}`)
    .reverse()
    .replace(new RegExp(`^\\${thousands}`, 'g'), '')
    .concat(`${decimal}${right}`)
  if (!currency) {
    return answer
  }
  return `${currency}${answer}`
}

/**
 * @param value
 * @param {string} thousands
 * @param {string} decimal
 * @param {string} currency
 * @returns {number | any}
 */
export const unMoney = (
  value: string,
  thousands: string = '.',
  decimal: string = ',',
  currency: string = 'R$ '
): number => {
  // noinspection RegExpRedundantEscape
  const string = String(value)
    .replace(currency, '')
    .replace(new RegExp(`\\${thousands}`, 'g'), '')
    .replace(new RegExp(`\\${decimal}`, 'g'), '.')
  return Number(string)
}
