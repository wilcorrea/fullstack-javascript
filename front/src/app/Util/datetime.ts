import dayJS from 'dayjs'

/**
 * @param {string} date
 * @param {string} format
 * @returns {string}
 */
export const format = (date: string, format: string = 'DD/MM/YYYY'): string => {
  return dayJS(date).format(format)
}
