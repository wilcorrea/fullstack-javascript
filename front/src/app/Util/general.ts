import $router from 'src/router'
import { replacement } from './string'

/**
 * @param {Object} element
 * @param {String|Array} path
 * @param {any} fallback
 * @returns {*}
 */
export const get = (element: any, path: any, fallback?: any): any => {
  if (element === undefined || element === null) {
    return fallback
  }

  const search = Array.isArray(path)
    ? path
    : path.split('.').filter((pieces: any) => pieces && pieces.length)

  if (!search.length) {
    return element
  }

  let property = search.shift()
  if (Array.isArray(element)) {
    // eslint-disable-next-line no-useless-escape
    property = String(property).replace(/[\[\]]+/g, '')
  }
  return get(element[property], search, fallback)
}

/**
 * @param target
 * @param options
 */
export const browse = (target: any, options: any = undefined) => {
  if (typeof target === 'string' && options && options.blank) {
    window.location.assign(target)
    return
  }

  if (typeof target === 'number') {
    $router.go(target)
    return
  }

  if (typeof target === 'string') {
    target = { path: target, query: {} }
  }

  if (options === true || (options && typeof options === 'object' && options.keep)) {
    target.query = { ...target.query, ...$router.currentRoute.query }
  }

  const regex = (expression: string) => new RegExp(`:${expression}`, 'g')
  target.path = replacement(target.path, $router.currentRoute.params, regex)

  $router.push(target)
}

/**
 * @param {*} element
 * @param {Function} action
 * @returns {*}
 */
export const clone = (element: any, action: Function = (value: any) => value): any => {
  // Handle the 3 simple types, and null or undefined
  if (element === null || element === undefined || typeof element !== 'object') {
    return action(element)
  }

  // Handle File
  if (element instanceof File) {
    return new File([element], element.name)
  }

  // Handle Date
  if (element instanceof Date) {
    const date = new Date()
    date.setTime(element.getTime())
    return action(date)
  }

  // Handle Array
  if (element instanceof Array) {
    return element.map((item) => clone(item, action))
  }

  // Handle Object
  if (element instanceof Object) {
    const reduce = (accumulate: any, property: string) => {
      accumulate[property] = clone(element[property], action)
      return accumulate
    }
    return Object.keys(element).reduce(reduce, {})
  }

  throw new Error('Unable to copy element! Its type isn\'t supported.')
}

/**
 * @returns {string}
 */
export const uuid = (): string => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1)
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4()
}
