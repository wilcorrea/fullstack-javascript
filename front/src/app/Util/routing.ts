import { RouteConfig } from 'vue-router'

/**
 * @param {string} path
 * @param {string} redirect
 * @returns RouteConfig
 */
export const redirect = (path: string, redirect: string): RouteConfig => {
  return { path, redirect }
}

/**
 * @param {string} path
 * @param {string} name
 * @param {Function} component
 * @param {any} [meta]
 * @returns RouteConfig
 */
export const route = (
  path: string,
  name: string,
  component: Function,
  meta: any = {}
): RouteConfig => {
  return { path, name, component, meta }
}

/**
 * @param {string} path
 * @param {Function} component
 * @param {Array<RouteConfig>>} [children]
 * @returns RouteConfig
 */
export const group = (
  path: string,
  component: Function,
  children: Array<RouteConfig> = []
): RouteConfig => {
  return { path, component, children }
}
