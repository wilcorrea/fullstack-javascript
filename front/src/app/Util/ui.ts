import Vue from 'vue'

/**
 * @param {string} name
 * @param {any} component
 */
export const register = (name: string, component: any): void => {
  Vue.component(name, component)
}

/**
 * @param {Element} element
 * @param {number} duration
 */
export const scrollUp = (element: Element, duration: number = 256) => {
  if (!element || typeof element !== 'object') {
    return
  }
  let counter = element.scrollTop
  scroller(element, (options: any) => {
    options.scrollCount += Math.PI / (duration / (options.newTimestamp - options.oldTimestamp))
    if (options.scrollCount >= Math.PI) {
      element.scrollTo(0, 0)
      return
    }
    if (counter <= 0) {
      return
    }
    counter = counter - duration / 10
    return Math.round(counter)
  })
}

/**
 * @param {Element} element
 * @param {number} duration
 */
export const scrollDown = (element: Element, duration: number = 256) => {
  if (!element || typeof element !== 'object') {
    return
  }
  const max = element.scrollHeight
  let counter = element.scrollTop
  scroller(element, (options: any) => {
    options.scrollCount += Math.PI / (duration / (options.newTimestamp - options.oldTimestamp))
    if (options.scrollCount >= Math.PI) {
      element.scrollTo(0, max)
      return
    }
    if (counter >= max) {
      return
    }
    counter = counter + duration / 10
    return Math.round(counter)
  })
}

/**
 * @param {Element} element
 * @param {Function} factor
 */
export const scroller = (element: Element, factor: Function) => {
  const options = {
    scrollCount: 0,
    oldTimestamp: performance.now(),
    newTimestamp: 0
  }

  const step = (newTimestamp: number) => {
    options.newTimestamp = newTimestamp
    const value = factor(options)
    if (!value) {
      return
    }
    element.scrollTo(0, value)
    options.oldTimestamp = newTimestamp
    window.requestAnimationFrame(step)
  }

  window.requestAnimationFrame(step)
}

/**
 * @returns {string}
 */
export const color = (): string => {
  const letters = '0123456789ABCDEF'
  const matches = []
  for (let i = 0; i < 6; i++) {
    matches.push(letters[Math.floor(Math.random() * 16)])
  }
  return `#${matches.join('')}`
}
