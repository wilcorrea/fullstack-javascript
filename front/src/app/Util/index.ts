import { get, clone } from './general'

export default {
  get, clone
}
