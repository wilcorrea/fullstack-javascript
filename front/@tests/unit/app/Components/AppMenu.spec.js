import { shallowMount } from '@vue/test-utils'
import AppMenu from 'src/app/Components/Menu/AppMenu.vue'
import AppIcon from 'src/app/Components/Icon/AppIcon'

import { register } from 'src/app/Util/ui'

register('AppIcon', AppIcon)

const label = 'Test'
register('RouterLink', { template: `<a>${label}</a>` })

describe('AppMenu', () => {
  const spy = jest.spyOn(console, 'error')

  /**
   * @test
   */
  it('"links" is required', () => {
    shallowMount(AppMenu)
    expect(spy).toBeCalledWith(expect.stringContaining('[Vue warn]: Missing required prop: "links"'))
  })

  const links = [{ path: '/test', label }]
  const wrapper = shallowMount(AppMenu, { propsData: { links } })

  /**
   * @test
   */
  it('create a simple menu', () => {
    expect(wrapper.find('ul > li > router-link-stub').text()).toMatch(label)
  })

  wrapper.vm.showMenu = true

  /**
   * @test
   */
  it('method appMenuToggle', () => {
    wrapper.vm.appMenuToggle()
    expect(false).toEqual(wrapper.vm.showMenu)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.vm.$el).toMatchSnapshot('closed')
  })

  /**
   * @test
   */
  it('method appMenuToggle', () => {
    wrapper.vm.appMenuToggle()
    expect(true).toEqual(wrapper.vm.showMenu)
  })

  /**
   * @test
   */
  test('should render content correctly', () => {
    expect(wrapper.vm.$el).toMatchSnapshot('open')
  })
})
