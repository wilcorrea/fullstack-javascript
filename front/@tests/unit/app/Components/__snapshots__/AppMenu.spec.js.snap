// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppMenu should render content correctly: closed 1`] = `
<div
  class="AppMenu"
>
  <div
    class="AppMenu__backdrop"
  />
   
  <div
    class="AppMenu__trigger"
  />
   
  <div
    class="AppMenu__links"
  >
    <div
      class="AppMenu__closer"
    >
      <app-icon-stub
        color="white"
        name="close"
      />
    </div>
     
    <ul>
      <li>
        <router-link-stub
          to="/test"
        >
          
          Test
        
        </router-link-stub>
      </li>
    </ul>
  </div>
</div>
`;

exports[`AppMenu should render content correctly: open 1`] = `
<div
  class="AppMenu"
>
  <div
    class="AppMenu__backdrop AppMenu__trigger--visible"
  />
   
  <div
    class="AppMenu__trigger AppMenu__trigger--triggered"
  />
   
  <div
    class="AppMenu__links AppMenu__links--open"
  >
    <div
      class="AppMenu__closer"
    >
      <app-icon-stub
        color="white"
        name="close"
      />
    </div>
     
    <ul>
      <li>
        <router-link-stub
          to="/test"
        >
          
          Test
        
        </router-link-stub>
      </li>
    </ul>
  </div>
</div>
`;
